content/posts/building-images-with-img.md

```diff
@@ -0,0 +1,257 @@
+
+kubectl annotate serviceaccount --namespace img img \
+  iam.gke.io/gcp-service-account=img-gcr@melgin.iam.gserviceaccount.com
+
+# bind GSA to KSA
+gcloud iam service-accounts add-iam-policy-binding \
+  --role roles/iam.workloadIdentityUser \
+  --member "serviceAccount:melgin.svc.id.goog[img/img]" \
+  img-gcr@melgin.iam.gserviceaccount.com
+```
+
+With our service accounts configured both in Google Cloud and our Kubernetes cluster, we can now create our `PodSecurityPolicy`:
+```yaml
+# based on https://github.com/kypseli/cb-core-oc-workshop/blob/master/k8s/cb-core-psp.yml
+apiVersion: policy/v1beta1
+kind: PodSecurityPolicy
+metadata:
+  name: img
+  namespace: img
+  annotations:
+    seccomp.security.alpha.kubernetes.io/allowedProfileNames: 'unconfined,runtime/default'
+    apparmor.security.beta.kubernetes.io/allowedProfileNames: 'unconfined,runtime/default'
+    seccomp.security.alpha.kubernetes.io/defaultProfileName:  'runtime/default'
+    apparmor.security.beta.kubernetes.io/defaultProfileName:  'runtime/default'
+spec:
+  allowPrivilegeEscalation: true
+  allowedProcMountTypes:
+  - Unmasked
+  - Default
+  fsGroup:
+    rule: 'MustRunAs'
+    ranges:
+      - min: 1
+        max: 65535
+  hostPID: false
+  hostIPC: false
+  hostNetwork: false
+  privileged: false
+  runAsUser:
+    rule: 'MustRunAs'
+    ranges:
+      - min: 1
+        max: 65535
+  seLinux:
+    rule: RunAsAny
+  supplementalGroups:
+    rule: RunAsAny
+  volumes:
+  - 'emptyDir'
+  - 'secret'
+  - 'downwardAPI'
+  - 'configMap'
+  - 'persistentVolumeClaim'
+  - 'projected'
+```
+
+There are a few details worth calling out in this policy:
+1. We use the `runAsUser` specification to ensure our `img` Pod runs as a non-zero/non-root user.
+2. `privileged` is set to `false`. However, it's worth noting that `allowPrivilegeEscalation` must be set to true for `img` to execute commands properly.
+3. For `img` to run properly, both `seccomp` and `AppArmor` profiles must be set to `unconfined`, which is [required by `runc`](https://github.com/genuinetools/img#running-with-docker).
+4. We include `allowedProcMountTypes` with both `Unmasked` and `Default` as accepted values - more on this in a minute.
+
+Next, we're going to attach our `img` `ServiceAccount` to a corresponding `Role` with a `RoleBinding`. This will allow our `img` `Pod` to use the custom `PodSecurityPolicy` we just created.
+```yaml
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: img
+rules:
+- apiGroups: ['extensions']
+  resources: ['podsecuritypolicies']
+  verbs:     ['use']
+  resourceNames:
+  - img
+- apiGroups: [""]
+  resources: ["pods"]
+  verbs: ["create","delete","get","list","patch","update","watch"]
+
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: RoleBinding
+metadata:
+  name: img
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: img
+subjects:
+- kind: ServiceAccount
+  name: img
+```
+
+Finally, here is our `Pod` definition:
```

Comment:
Wouldn't be better to say `PodTemplate` definition instead of `Pod` definition? In that case it is understood that you are using this PodTemplate en the Jenkinsfile when you call the imgPod.yaml.

Reply:
Good point - I added some clarification that we'd ultimately be using this as a `PodTemplate`.